rfcs/027-pipeline-intermediate-storage/README.md

```diff
@@ -0,0 +1,147 @@
+
+* Transaction support might be useful for some of our current logic or future use cases.
+
+Cons:
+
+* Performance may not match other options due to stronger consistency guarantees (though likely to be much better than VHS due to only hitting a single system and making use of a cache).
+
+* Writing raw SQL is not fun, and I have not really been satisfied with any of the Scala wrappers I have tried.
+
+* Visibility nothing like Kibana (there are some web interfaces available which attempt to do similar things, but would require more manual setup and maybe not as polished).
+
+Questions:
+
+* Costs?
+
+* Will not scale as well to mega large data sets: probably not an issue with the sizes of data we are likely to deal with, but worth considering.
+
+## Implementation
+
+Orthogonal to the choice of store, the way we write data into the store for usage later in the pipeline has two main possibilities.
+
+### Service
+
+The current `recorder` service is a wrapper round VHS which stores items received at its input, emitting a message contained the ID of the stored item for downstream pipeline stages to use themselves for access in future.
+
+ We could mirror this design using a different store to VHS for upcoming intermediate storage needs. As the behaviour would be essentially the same regardless of what state the works were in (unmerged, merged, enriched etc) the exact same service can be inserted in different parts of the pipeline differing only in configuration of what store the service has access to.
+
+### Module
+
+An alternative to using a separate service would be to have the services performing the data transformation writing to the store directly themselves, such as through a common module (`scala-storage` may already serve this purpose sufficiently). This would clearly result in less infrastructure, and so would reduce costs and possibly maintenance burden.
+
+The main issue with this approach is that there is potentially less delineation of concerns, with stages of the pipeline performing work additional to solely data transformation. Any failures during the storage would result in the whole incoming message being marked as a failure, with any previous work on transforming the data having to be redone. Having a separate service solely for storage thus means the point of failure for a particular message making its way through the pipeline is more granular.
+
+However, it is worth mentioning that the waters are muddied somewhat due to the fact that the ability to receive whole works as input to some service requires the use of `big_messaging`. In the case of works over 256KB these works will be stored twice, once in S3 and subsequently in the intermediate store itself. Beyond just reducing costs related to this repeated storage, it may make the whole architecture more easily understandable and improve visibility if we just used intermediate stores and the passing of keys between services, rather than requiring both systems.
```

Comment:
>if we just used intermediate stores and the passing of keys between services, rather than requiring both systems.
👍

Reply:
I'm not sure about this idea - if we don't always have intermediate stores between services then I think it will get complicated: a shared database across many microservices is generally regarded as a Bad Idea.

Another way of saying this: not sure that bigmessaging and intermediate storage are the same problem / are solved by the same solution. Intermediate storage is required for presenting a consistent view of data to multiple services, whereas bigmessaging is about pushing large bits of transformed data between services.

Reply:
The anti pattern IMO is when there is multiple services writing into the same database, I do not think having single write with one or two places reading out (like we do with `recorder` and `matcher` / `merger`) is going to be problematic. I am not suggesting one single store, but a few stores at different points in the pipeline (e.g. the relation embedder store does not know anything about any other stores apart from the merged works store, so still encapsulation in that sense).

`big_messaging` and intermediate storage shouldn't be the same problem I agree, but due to limitations in SQS it complicates things. Worth mentioning also that SQS could also be viewed as a form of intermediate storage (a queue, rather than key / value store).

Reply:
Also worth mentioning that we already pass keys into databases between services, this is not something new. The choice isn't really between big bits of data or passing keys or, it's between passing big bits of data and then keys, or just passing the keys.

Reply:
Sorry, I was only referring to the idea of bigmessaging using an intermediate store like this - I agree that there is no problem with the proposed pattern of single write / few read stores, but I don't think we should be extending that to bigmessaging.

Reply:
Ah ok, makes sense :+1:

Reply:
I still _think_ BigMessaging could be made redundant to simplify the pipeline with the intermediary storage solution, if we were happy that every step in the pipeline emitted IDs and proceeding step did a lookup.

I am pretty certain we would hit bottlenecks while reindexing, which is something to think about irrespective of ☝️ - if these stores are going to written to what is the throughput we need to support, although the ingestors are a good example and _seem_ to deal with the load.

Reply:
Yes I agree big messaging should be removed where not necessary. I don't see how we would introduce new bottlenecks into the pipeline that don't exist already, assuming any new store we use is at least as fast as VHS. The relation embedder could introduce a new bottleneck but that will more be to do with the requirements of that service rather than using intermediate storage vs big messaging.